{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efdd5288-af92-404d-b3fc-bda85288ea0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5c7d3eea-fa77-48ff-9468-e54ea1caf609",
   "metadata": {},
   "source": [
    "# GWAS Lab\n",
    "\n",
    "In this lab we will be running a GWAS with data from *Arabidopsis thaliana*.\n",
    "\n",
    "## The Data\n",
    "\n",
    "The data for this lab comes from [The 1001 Genomes Project](https://1001genomes.org/), specifically the [EasyGWAS website](https://easygwas.biochem.mpg.de/down/1/). \n",
    "\n",
    "The genomic data are in three files in /mnt/research/PLB812_FS24_S001/08_GWAS_Josephs/\n",
    "\n",
    "`arabidopsis.bed` has the genotypes for every individual in the dataset. <br>\n",
    "`arabidopsis.bim` has the locations of every SNP in `genotype.ped` <br>\n",
    "`arabidopsis.fam` tells you the order of the genotypes in the file.\n",
    "\n",
    "The phenotypic data is in `phenotypes.pheno`.\n",
    "\n",
    "\n",
    "## Installing GEMMA\n",
    "\n",
    "We will be using the software [GEMMA](https://github.com/genetics-statistics/GEMMA) to run the GWAS. The HPCC does not maintain GEMMA, but it is pretty easy to get running on your own. \n",
    "\n",
    "First, use wget to download GEMMA. I find it easiest to make an \"Apps\" folder in my home directory on the HPCC that I use to keep the programs I download, but it doesn't actually matter where you download this file.\n",
    "\n",
    "`$ wget https://github.com/genetics-statistics/GEMMA/releases/download/v0.98.5/gemma-0.98.5-linux-static-AMD64.gz`\n",
    "\n",
    "Next, use gunzip to uncompress the file.\n",
    "\n",
    "`$ gunzip gemma-0.98.5-linux-static-AMD64.gz`\n",
    "\n",
    "Use chmod to make this file executable. \n",
    "\n",
    "`$chmod u+x gemma-0.98.5-linux-static-AMD64`\n",
    "\n",
    "Now, run the file to check to see what happens\n",
    "\n",
    "`$ ./gemma-0.98.5-linux-static-AMD64`\n",
    "\n",
    "If everything is working right you should get an introductory message about the program.\n",
    "\n",
    "## Running a simple GWAS\n",
    "\n",
    "### Making the phenotype file (Emily has already done this for you but I left the code here in case you need it).\n",
    "\n",
    "We need to add our phenotypes to the **.fam** file. We'll be looking at the trait **DTF1** which describes days until bolting. This is an important reproductive trait. This is the fourth trait in the phenotype file **phenotypes.pheno**.\n",
    "\n",
    "First, rename the `arabidopsis.fam` file so you have it for later.\n",
    "\n",
    "`$mv arabidopsis.fam arabidopsis-old.fam`\n",
    "\n",
    "Below is some code to do this. Note that you will need to change the paths to the fam, pheno, and out files.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a200361b-d390-4f00-85d3-c16876a23dfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#open the files\n",
    "famFile = open(\"/mnt/home/josep993/plb812/gwas-data/arabidopsis-old.fam\",'r')\n",
    "phenoFile = open('/mnt/home/josep993/plb812/gwas-data/phenotypes.pheno','r')\n",
    "\n",
    "#make a dictionary for the phenotype data\n",
    "phenoDict = {}\n",
    "\n",
    "#read in the phenotype data\n",
    "phenos = phenoFile.readlines()\n",
    "for phen in phenos:\n",
    "    p = str.split(phen)\n",
    "    dtf=p[8]\n",
    "    if (dtf=='nan'):\n",
    "        dtf=\"-9\"\n",
    "    #change the NA strings\n",
    "    #add to the dictionary where key is the genotype and the value is the phenotype\n",
    "    phenoDict[p[1]]= dtf\n",
    "phenoFile.close()\n",
    "\n",
    "#open the output file\n",
    "outFile = open(\"/mnt/home/josep993/plb812/gwas-data/arabidopsis.fam\",\"w\")\n",
    "\n",
    "#read in the genotype file -- we want our output to be in this order!\n",
    "genos = famFile.readlines()\n",
    "for geno in genos:\n",
    "    g = str.split(geno)\n",
    "    #make the output line from the genotype file plus the phenotype\n",
    "    newg = [*g[0:5],phenoDict[g[1]],'\\n']\n",
    "    myout = '    '.join(newg)\n",
    "    outFile.write(myout)\n",
    "\n",
    "#close all the files\n",
    "famFile.close()\n",
    "outFile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fb685f8-84de-49db-902c-acf0102052ba",
   "metadata": {},
   "source": [
    "### Running the GWAS\n",
    "\n",
    "Here is the code for running the GWAS.\n",
    "\n",
    "`../../Apps/gemma-0.98.5-linux-static-AMD64 -bfile arabidopsis -lm 2 -n 1 -o dtf`\n",
    "\n",
    "You can read the manual for all the options, but note that the **-lm** flag sets the type of model to run, and the **-n** flag tells you which phenotype to use if the **.fam** file has multiple phenotypes.\n",
    "\n",
    "I'm not sure if HPCC will handle the whole class running GEMMA on the same development node at the same time, so you may want to put this code into a bash script and submit it to slurm. As a reminder, here is a guide to do this: https://docs.icer.msu.edu/Job_Script_and_Job_Submission/. Also, keep in mind that you can edit bash script files in ondemand.\n",
    "\n",
    "\n",
    "## Looking at the data\n",
    "\n",
    "I wrote a short R script to make a manhattan plot of the data.\n",
    "\n",
    "`$module load R-bundle-CRAN/2023.12-foss-2023a`\n",
    "\n",
    "`$Rscript --vanilla manplot.R [path to gemma output] [output file name]`\n",
    "\n",
    "The R script makes a manhattan plot and a qq plot in an .eps format. Download these plots and investigate them. \n",
    "\n",
    "## Running a mixed model GWAS\n",
    "\n",
    "Next, we will run a mixed model GWAS that accounts for relatedness between the Arabidopsis samples.\n",
    "\n",
    "\n",
    "### Making a kinship matrix\n",
    "\n",
    "This command makes a kinship matrix.\n",
    "\n",
    "`$../../Apps/gemma-0.98.5-linux-static-AMD64 -bfile arabidopsis -gk 2 -o arabidopsis`\n",
    "\n",
    "the **-gk** flag tells GEMMA that we want to make a standardized kinship matrix (instead of a centered one). You can read more in the manual about the distinction between these two options.\n",
    "\n",
    "### Running the GWAS\n",
    "\n",
    "This is the command for running the GWAS with our new kinship matrix. \n",
    "\n",
    "`$../../Apps/gemma-0.98.5-linux-static-AMD64 -bfile arabidopsis -k output/arabidopsis.sXX.txt -lmm 2 -n 1 -o dtf-mixed`\n",
    "\n",
    "We can make more plots with the R script.\n",
    "\n",
    "`$Rscript --vanilla manplot.R [path to gemma output] [output file name]`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaff619f-dd31-4f7d-b9ff-b56001e8fc22",
   "metadata": {},
   "source": [
    "### Questions \n",
    "\n",
    "1) Please add both of your manhattan plots and both of your qq plots to the ipython notebook that you submit.\n",
    "\n",
    "\n",
    "2) How do the Manhattan plots of the simple GWAS and the mixed model GWAS differ?\n",
    "   \n",
    "3) How do the QQ Plots of the simple GWAS and the mixed model GWAS differ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "094287b3-31c8-4f20-831c-2dc6bd18f34f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#1. I added my plots as a pdf because I did not know how to add them to my notebook. I was absent in class for this lab.\n",
    "#2. The manhattan plot of the simple gwas shows most (if not all) of the p-values for each snp are above the significance threshold. Conversely, the mixed model GWAS manhattanplot has muchb lower -log10(p) values and significance threshold ; only two loci have significat snps.\n",
    "#3. The QQ plot of the simple GWAS shows divergence of the p-values form the expected for the majority of the loci. On the other hand, the mixed model qqplot shows most of the loci adhere to the expected p-values. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
